File: cgi-bin/hello.php
<?php
$lengthRaw = $_SERVER['CONTENT_LENGTH'] ?? '0';
$bodySize = (int)$lengthRaw;

$body = '';
if ($bodySize > 0) {
    $body = file_get_contents('php://stdin', false, null, 0, $bodySize);
    if ($body === false) {
        $body = '';
    }
}

$query = $_SERVER['QUERY_STRING'] ?? null;

header('Content-Type: text/html; charset=UTF-8');

echo "<html>\n<body>\n";
echo "<h2>Hello world!</h2>\n";
echo "<p>Your php CGI is set up properly</p>\n";

if ($query !== null) {
    echo "<p>Here is your query string data: " .
         htmlspecialchars($query, ENT_QUOTES | ENT_SUBSTITUTE, 'UTF-8') .
         "</p>\n";
}

echo "<p>Here is your body string data: " .
     htmlspecialchars($body, ENT_QUOTES | ENT_SUBSTITUTE, 'UTF-8') .
     "</p>\n";

echo "</body>\n</html>\n";
?>

File: cgi-bin/hello.py
import os
import sys
import html

length_raw = os.getenv("CONTENT_LENGTH", "0")
try:
    body_size = int(length_raw)
except ValueError:
    body_size = 0

body = sys.stdin.read(max(body_size, 0))
query = os.getenv("QUERY_STRING")

# CGI scripts should emit headers first
# print("Content-Type: text/html")
print()

print("<html>")
print("<body>")
print("<h2>Hello world!</h2>")
print("<p>Your python CGI is set up properly</p>")

if query is not None:
    print(f"<p>Here is your query string data: {html.escape(query)}</p>")

print(f"<p>Here is your body string data: {html.escape(body)}</p>")
print("</body>")
print("</html>")
